### PR TITLE
Shadow model training

### PR DESCRIPTION
**Pull request by GalHorowitz**
# Description

The PR adds a utility for training shadow models and in turn using them to generate shadow-datasets for use in membership inference attacks. In addition, an implementation of the hill-climbing synthetic data generation algorithm (from R. Shokri et al. 2017) is added which makes it possible to train shadow models with no existing dataset.
A notebook demonstrating usage in a black-box membership inference attack was also added.
A setter for the `nb_classes` estimator parameter and usages of `nb_classes` have been updated so the `set_params` method used in estimator cloning can work correctly.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

All unit tests pass.

*Test Configuration*:
- OS: MacOS 11.4
- Python version: 3.8.11

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by GalHorowitz** [comment]
The two failing tests do not seem related to my changes, and pass locally.

**Comment by beat-buesser** [comment]
Hi @GalHorowitz Thank you very much! I have restarted the tests and changed the base branch to `dev_1.9.0`.

**Comment by beat-buesser** [comment]
I will merge this pull request to branch `shadow_model_training` and address the review comments there before merging it to `dev_1.9.0`.

File: art/attacks/inference/membership_inference/shadow_models.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/inference/membership_inference/shadow_models.py:3]: I agree.

File: art/attacks/inference/membership_inference/shadow_models.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_TYPE
+    from art.estimators.classification.scikitlearn import ScikitlearnClassifier
+    from art.estimators.classification import PyTorchClassifier, TensorFlowV2Classifier
+
+
+class ShadowModels:
+    """
+    Utility for training shadow models and generating shadow-datasets for membership inference attacks.
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/inference/membership_inference/shadow_models.py:39]: I agree.

File: art/attacks/inference/membership_inference/shadow_models.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    def _default_randomize_features(self, record: np.ndarray, num_features: int) -> np.ndarray:
+        new_record = record.copy()
+        for _ in range(num_features):
+            new_record[self._rng.integers(0, self._input_shape)] = self._rng.random()
+        return new_record
+
+    def _hill_climbing_synthesis(
+        self,
+        target_classifier: "CLASSIFIER_TYPE",
+        wanted_class: int,
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/inference/membership_inference/shadow_models.py:147]: I agree.

File: art/attacks/inference/membership_inference/shadow_models.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+            x = randomize_features_fn(best_x, k_features_randomized)  # type: ignore
+
+        raise RuntimeError("Failed to synthesize data record")
+
+    def generate_synthetic_shadow_dataset(
+        self,
+        target_classifier: "CLASSIFIER_TYPE",
+        dataset_size: int,
+        max_features_randomized: int,
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/inference/membership_inference/shadow_models.py:228]: I agree.

File: tests/attacks/inference/membership_inference/test_shadow_models.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        )
+        (mem_x, mem_y, mem_pred), (nonmem_x, nonmem_y, nonmem_pred) = shadow_dataset
+
+        assert len(mem_x) == len(mem_y)
+        assert len(mem_y) == len(mem_pred)
+        assert len(nonmem_x) == len(nonmem_y)
+        assert len(nonmem_y) == len(nonmem_pred)
+        assert len(mem_x) + len(mem_y) == 40
+    except ARTTestException as e:
+        art_warning(e)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply tests/attacks/inference/membership_inference/test_shadow_models.py:129]: I agree.

**Pull request merged**